src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java

```diff
@@ -67,21 +71,31 @@ public RetentionRulesServiceImpl() {
    * @param rule the {@link RetentionRuleCreateRequest} object input by the user
    * @return the {@link Integer} id of the created rule
    */
-  @Override()
+  @Override
   public Integer createRetentionRule(RetentionRuleCreateRequest rule, UserInfo user)
-      throws SQLException {
+      throws SQLException, IOException {
+    String bucketName = RetentionUtil.getBucketName(rule.getDataStorageName());
+    if (rule.getRetentionRuleType() != RetentionRuleType.GLOBAL) {
+      Bucket bucket = GcsHelper.getInstance().getBucket(bucketName);
+      if (bucket == null) {
+        throw new IOException(String.format("Bucket %s does not exist", bucketName));
+      }
+    }
     String userName = user.getEmail() == null ? DEFAULT_UNKNOWN_USER : user.getEmail();
 
-    RetentionRule existing =
-        ruleDao.findByBusinessKey(
-            rule.getProjectId(), rule.getDataStorageName(), true, rule.getRetentionRuleType());
+    List<RetentionRule> bucketRules =
+        ruleDao.findRulesByStorageRoot(
+            rule.getProjectId(),
+            RetentionUtil.getBucketName(rule.getDataStorageName()),
+            rule.getRetentionRuleType(),
+            true);
 
-    RetentionRule newRule;
+    RetentionRule existing = getExistingRule(bucketRules, rule);
```

Comment:
consider renaming to existingRule for increased readability

Reply:
done